src/net_processing.cpp

```diff
@@ -821,7 +836,21 @@ void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTim
     g_outbound_peers_with_protect_from_disconnect -= state->m_chain_sync.m_protect;
     assert(g_outbound_peers_with_protect_from_disconnect >= 0);
 
-    g_outbound_peers.erase(nodeid);
+    // Remove entry from g_outbound_peers, if present.
+    size_t peer_loc = g_outbound_peers.size();
+    for (size_t i=0; i<g_outbound_peers.size(); ++i) {
+        if (g_outbound_peers[i] == nodeid) {
+            peer_loc = i;
+        }
+    }
+    if (peer_loc != g_outbound_peers.size()) {
+        // We found the peer's location in the outbound peer set
+        if (peer_loc != g_outbound_peers.size()-1) {
+            // Swap this peer with the last one
+            g_outbound_peers[peer_loc] = g_outbound_peers.back();
+        }
+        g_outbound_peers.pop_back();
+    }
```

Comment:
Seems like this could be a lot more succinct with [`vector::erase`](https://en.cppreference.com/w/cpp/container/vector/erase) - is there some reason to do it this way?

Reply:
I guess I have an aversion to vector::erase because it requires moving all the other entries.  Possibly that concern doesn't make sense for such a small vector?

But the bigger issue here is why this is a vector at all -- originally I had this as a set, but then I changed it to be a vector in order to make the logic around when to drain the notfound queue for each peer make sense.  In short, I wanted to uniformly at random assign a notfound transaction to an outbound peer that announced it, but I didn't have a good way to do that, so instead I assign it to all peers, but then have a (hacky) counter in SendMessages and each outbound peer will check that counter to decide when to drain its notfound queue.

(I'm hoping that reviewers will give me a better idea of how to implement that logic than what I've got so far...)